torch/csrc/autograd/python_variable.cpp

```diff
@@ -432,6 +433,20 @@ static PyObject * THPVariable_layout(THPVariable* self, PyObject* args) {
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject * THPVariable_device(THPVariable* self, PyObject* args) {
+  HANDLE_TH_ERRORS
+  auto& self_ = self->cdata;
+  std::ostringstream oss;
+  if (self_.type().is_cuda()) {
+    oss << "cuda:" << self_.get_device();
+  }
+  else {
+    oss << "cpu";
+  }
+  return THPUtils_packString(oss.str().c_str());
```

Comment:
Why don't we return a `DeviceSpec` here?

Reply:
For one, that's what tf eager does.  I also thought most of the time people would just pass around strings, and would only use DeviceSpecs when they wanted to avoid parsing and/or to do math on the indices.

But I'm fine having a `device` and a `device_str`, that might just be the easiest.

Reply:
Yes, I think `device_str` is better (although it's just `str(x.device)`, so not sure if we should have it).

Returning a `DeviceSpec` has the benefit that:
1. If you want to get more detailed info (e.g. index), you don't need to parse the string again
2. If you pass it to any of our methods, it's cheaper to arg check and ignore string parsing at every call.

I think we should discourage people from passing around strings, or they will accidentally start slowing down their code considerably.